### PR TITLE
FT: Add lifecycle producer

### PR DESCRIPTION
**Pull request by bennettbuchanan**

### CODE REVIEW

File: extensions/replication/queueProcessor/task.js [hunk 1]
```diff
@@ -1,6 +1,5 @@
 'use strict'; // eslint-disable-line
 const assert = require('assert');
-const werelogs = require('werelogs');
```
Comment (jonathan-gramain) [inline extensions/replication/queueProcessor/task.js]: Was this change intended to be committed?

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+     * @param {number} authConfig.vault.adminPort - vault admin port
+     * @param {String} transport - http or https
+     */
+    constructor(zkConfig, lcConfig, s3Config, authConfig, transport) {
+        this._log = new Logger('Backbeat:LifecycleProducer');
+        this._zkConfig = zkConfig;
+        this._lcConfig = lcConfig;
+        this._s3Client = new AWS.S3({
+            endpoint: `${transport}://${s3Config.host}:${s3Config.port}`,
+            credentials: new AccountCredentials(authConfig, this._log),
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: When using vault, we will have to use the admin credentials in admin-backbeat.json to create new access keys for the account owning the bucket, and cache them in memory, like echo feature is doing (it also creates accounts on the destination but we do not want this for lifecycle). See how the `VaultClientCache` class is used to do this.

A corollary is that we'll need different S3 clients for each account, each time providing them the proper account access/secret key for the object/bucket to process. For replication, we create new clients for every object to replicate, sounds a bit overkill but does not seem to cause major CPU/memory issues in practice so I think it's acceptable.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            this._log.error('could not parse bucket entry', { error });
+            return process.nextTick(() => cb(error));
+        }
+        return this._s3Client.getBucketLifecycleConfiguration({
+            Bucket: value.bucket,
+        }, (err, config) => {
+            if (err) {
+                this._log.error('error getting bucket lifecycle configuration',
+                    { error: err });
+                return process.nextTick(() => cb(err));
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: since we're already in an async context (answer from `getBucketLifecycleConfiguration) we can get rid of the `process.nextTick`

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+            }
+            if (!this._shouldProcessConfig(config)) {
+                return process.nextTick(cb);
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: same here

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+class LifecycleProducer {
+    /**
+     * Constructor of LifecycleProducer
+     * @param {Object} zkConfig - zookeeper config
+     * @param {Object} lcConfig - lifecycle config
+     * @param {Object} s3Config - s3 config
+     * @param {String} s3Config.host - host ip
+     * @param {String} s3Config.port - port
+     * @param {Object} authConfig - auth config
+     * @param {String} authConfig.account - account name
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Technically, only one of either `account` or `vault` sections is mandatory, so you can specify both as optional.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+    /**
+     * Send entry to the object task topic
+     * @param {[type]} entry - The Kafka entry to send to the topic
+     * @param {Function} cb - The callback to call
+     * @return {undefined}
+     */
+    sendObjectEntry(entry, cb) {
+        const strEntry = JSON.stringify(entry);
+        this._objectProducer.send([strEntry], cb);
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: I think this will not work, because the `send` method expects an array of *objects*, where each object has a `message` attribute and optionally a `key` attribute as well. So normally the proper way would be to construct an array of objects with a `message` attribute which is the result of `JSON.stringify`, if I'm not mistaken.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+    /**
+     * Send entry back to bucket task topic
+     * @param {[type]} entry - The Kafka entry to send to the topic
+     * @param {Function} cb - The callback to call
+     * @return {undefined}
+     */
+    sendBucketEntry(entry, cb) {
+        const str = JSON.stringify(entry);
+        this._bucketProducer.send([str], cb);
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: same issue here

File: extensions/replication/queueProcessor/task.js (hunk 1 above)
Comment (bennettbuchanan) [reply extensions/replication/queueProcessor/task.js]: Yes, I noticed it was not used in the file. I guess maybe it is outside the scope of this PR though.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 6 above)
Comment (philipyoo) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: So just making sure, but would it then be something like `[ { message: strEntry } ]`

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 5 above)
Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Updated.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 4 above)
Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Updated.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 3 above)
Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Updated.

File: extensions/replication/queueProcessor/task.js (hunk 1 above)
Comment (bennettbuchanan) [reply extensions/replication/queueProcessor/task.js]: Removed these changes.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 6 above)
Comment (jonathan-gramain) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: yes

Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Updated.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 7 above)
Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Updated.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 2 above)
Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: I've updated this to generate a new key for each unique canonical ID (using cache if the canonical ID has been seen before). I then create a new S3 client for each `getBucketLifecycle` request using the access/secret keys.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+        consumer.subscribe();
+    }
+
+    /**
+     * Set up the vault client cache for making requests to vault.
+     * @return {undefined}
+     */
+    _setupVaultClientCache() {
+        const { vault } = this._authConfig;
+        const { host, port, adminPort, adminCredentialsFile } = vault;
```
Comment (philipyoo) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: just a quick note, `adminCredentialsFile` will need to be added to config files and a joi required field

Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Was relying on this addition: https://github.com/scality/backbeat/blob/ft/S3C-1154-lifecycle/conf/config.joi.js#L31 and the error thrown if we cannot correctly parse credentials from the file. I guess we should also check for the existence of the `adminCredentialsFile` field, or else change `.optional()` to `.required()` in the joi validation.

**Comment by bennettbuchanan** [comment]
@LaurenSpiegel brought up a good point that this class handles more than just setting up producers for Kafka—it also sets up consumers, does a listing and filtering of the lifecycle config, and manages passing the config to the lifecycle task. 

What do you all think about updating the name of this component to "LifecycleProcessor" to follow the pattern of the replication extension? This differs from the original design doc so wanted to get everyone's input. @jonathan-gramain @dora-korpar @philipyoo

**Comment by dora-korpar** [comment]
That makes sense to me

**Comment by philipyoo** [comment]
Yea that sounds good to me 👍

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+     * @param {Function} cb - The callback to call
+     * @return {undefined}
+     */
+    _processBucketEntry(entry, cb) {
+        const { error, result } = safeJsonParse(entry.value);
+        if (error) {
+            this._log.error('could not parse bucket entry', { error });
+            return process.nextTick(() => cb(error));
+        }
+        const { bucket, owner } = result.value;
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: I think the correct deconstruction is `const { bucket, owner } = result;`, because the kafka message object has no "value" field but "bucket" and "owner" are top-level fields of the message object.

**Comment by jonathan-gramain** [comment]
OK for name change to lifecycle processor, don't forget to update also the config.json section (and will have to do it in Federation PR as well).

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 9 above)
Comment (jonathan-gramain) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: This exception occurred:
```
/home/scality/backbeat/extensions/lifecycle/lifecycleProducer/LifecycleProducer.js:177
        const { bucket, owner } = result.value;
                                        ^

TypeError: Cannot match against 'undefined' or 'null'.
    at LifecycleProducer._processBucketEntry (/home/scality/backbeat/extensions/lifecycle/lifecycleProducer/LifecycleProducer.js:177:41)
    at /home/scality/backbeat/node_modules/async/dist/async.js:4082:9
    at Object.process (/home/scality/backbeat/node_modules/async/dist/async.js:2330:17)
    at /home/scality/backbeat/node_modules/async/dist/async.js:2238:19
    at Immediate.<anonymous> (/home/scality/backbeat/node_modules/async/dist/async.js:119:16)
    at runCallback (timers.js:649:20)
    at tryOnImmediate (timers.js:622:5)
    at processImmediate [as _immediateCallback] (timers.js:594:5)
```

We should both fix the case, and also make the parsing of messages more resilient (i.e. do not crash but log an error and keep processing the next messages).

Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: I had seen this structure when publishing to a topic with a Kafka producer and needed to deconstruct this way. I will validate this one way or another as I'm adding tests.

Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Okay, I will update accordingly.

Comment (philipyoo) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: I think Bennett added a good try/catch for json parsing. For now, with the fix, I also check to make sure `bucket` and `owner` exist. If not, log error and return

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 10]
```diff
@@ -173,7 +173,15 @@ class LifecycleProducer {
             this._log.error('could not parse bucket entry', { error });
             return process.nextTick(() => cb(error));
         }
-        const { bucket, owner } = result.value;
+        const { bucket, owner } = result;
+        if (!bucket || !owner) {
+            this._log.error('kafka bucket entry missing required fields', {
```
Comment (jonathan-gramain) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: :+1:

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+        // The task scheduler for processing lifecycle tasks concurrently.
+        this._internalTaskScheduler = async.queue((ctx, cb) => {
+            const { task, rules, value } = ctx;
+            return task.processBucketEntry(rules, value, cb);
+        }, this._lcConfig.producer.concurrency);
+
+        // Listen for errors from any task being processed.
+        this._internalTaskScheduler.drain(err => {
+            if (err) {
+                this._log.error('error occured when during task processing', {
```
Comment (dora-korpar) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: trivial, but typo including `when during`. should use either one or the other

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+        return {
+            sendBucketEntry: this.sendBucketEntry,
+            sendObjectEntry: this.sendObjectEntry,
+            enabledRules: this._lcConfig.rules,
+            log: this._log,
+        };
+    }
+
+    /**
+     * Return an S3 client instance using the given accound credentials.
```
Comment (dora-korpar) [inline extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: ~~acound~~ -> account

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 11 above)
Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Updated.

File: extensions/lifecycle/lifecycleProducer/LifecycleProducer.js (hunk 12 above)
Comment (bennettbuchanan) [reply extensions/lifecycle/lifecycleProducer/LifecycleProducer.js]: Updated.

**Comment by jonathan-gramain** [comment]
@ironman-machine try

**Comment by jonathan-gramain** [comment]
@ironman-machine

**Comment by jonathan-gramain** [comment]
@ironman-machine 
DEFAULT_BRANCH=rel/7.4
try

**Comment by jonathan-gramain** [comment]
@ironman-machine 
DEFAULT_BRANCH=rel/7.4
SCALITY_INTEGRATION_BRANCH=ultron/rel/7.4
try

**Comment by jonathan-gramain** [comment]
@ironman-machine 
SCALITY_FEDERATION_BRANCH=ft/S3C-1154-lifecycleDeployment
SCALITY_METADATA_BRANCH=zenko/bbp2-rc1
SCALITY_INTEGRATION_BRANCH=ultron/rel/7.4
DEFAULT_BRANCH=rel/7.4
try

**Comment by jonathan-gramain** [comment]
@ironman-machine
SCALITY_FEDERATION_BRANCH=ft/S3C-1154-lifecycle
SCALITY_METADATA_BRANCH=zenko/bbp2-rc1
SCALITY_INTEGRATION_BRANCH=ultron/rel/7.4
DEFAULT_BRANCH=rel/7.4
try
